### PR TITLE
Add ProblemSolver with pruning and memoization

### PR DESCRIPTION
**Pull request by zbraniecki**
Ok, this is I think ready for a big-picture review.

I introduce the solver with two modes - serial and parallel. I refactor the internals to make it use the solver.
You can test it via:

```
cargo test --features tokio
cargo bench --features tokio
```

What I don't like about what I have:

* I'm not certain if it's true, but I sense that we ended up with too many structs in total between various levels of L10nRegistry and ProblemSolver, Solution and generators, streams and iterators.
* I don't like the duplication between `next` and `next_poll`/`next_bundle` in Serial/Parallel ProblemSolver. The streaming one returning `&Vec<usize>` is nice for testing and separation of concerns, but it all gets lost when I have to duplicate the loop for the one that returns a bundle. On the other hand, if I try to just keep `Vec<usize>` in Solver, and make it use the solution to build the bundle in the caller, I end up conflating use of the solver cache. I really would like to end up with solver having one method, and preferably only one returning `&[usize]`.
* I'm not sure if the solution/solver separation makes sense anymore
* parallel is slower - is it fixable?
* I'm cloning in so many places...

I'd like to keep fixing the issues, but I think I exhausted my capability to improve the code and I need @djg with fresh mind to look at my code and refactor it further.
I think the general solution holds, but I think that the architecture can be cleaned up and I'm not the right person to notice how because I wrote it.

@djg - can you take the next stab at putting things in place?

### CODE REVIEW

File: src/solver/README.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
+
+Source Order Problem Solver
+======================
+
+This module contains an algorithm used to power the `FluentBundle` generator in `L10nRegistry`.
+
+The main concept behind it is a problem solver which takes a list of resources and a list of sources and computes all possible iterations of valid sets of orders of sources that allow for creation of `FluentBundle` with the requested resources.
```
Comment (dminor) [inline src/solver/README.md]: I found this sentence difficult. Could you say "all valid combinations of source orders that allow..."?

File: src/solver/README.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    .multi_cartesian_product();
+
+for set in multi_prod {
+    println!("{:?}", set);
+}
+```
+([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6ef231f6b011b234babb0aa3e68b78ab))
+
+#### Python
+In Python, `itertools` package provides a function [`itertools::product`](https://docs.python.org/2/library/itertools.html#itertools.product) which can be used to generate such iterator:
```
Comment (dminor) [inline src/solver/README.md]: nit: We should probably link to the python3 docs, https://docs.python.org/3/library/itertools.html#itertools.product.

File: src/solver/README.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+```
+###### Diagrams generated with use of http://marklodato.github.io/js-boxdrawing/
+
+where the horizontal block is a candidate, vertical block is a set of sources possible for each resource, and the arrow represents the index of a resource the iterator is currently evaluating.
+
+With those tools introduced, we can now guide the reader through how the algorithm works.
+But before we do that, it is important to justify writing a custom algorithm in place of existing generic solutions, and explain the two testing strategies which heavily impact the algorithm.
+
+# Existing libraries
+The reader may notice that the concept presented here is known as Cartesian Multi-Product There are ready to use solutions for generating such iterators.
```
Comment (dminor) [inline src/solver/README.md]: I struggled with this a bit, because I'd never heard of the Cartesian Multi-Product before. I am familiar with the plain old Cartesian Product. I'd suggest explaining in terms of the Cartesian Product first, and then saying it can be represented more concisely using the multi-product. You could move the Python example first, since the product there takes multiple iterators, e.g.
```
for s in itertools.product(range(2), range(2), range(2)):
    print(s)
```
and then say it can be represented concisely using the Rust multi-cartesian product or the repeat argument in Python. Linking to the wikipedia page for Cartesian Product (https://en.wikipedia.org/wiki/Cartesian_product) wouldn't hurt.

File: src/solver/README.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+for set in itertools.product(range(2), repeat=3):
+    print(set)
+```
+
+#### Reasons for a custom algorithm
+
+Unfortunately, the computational complexity of generating all possible sets is growing exponentially, both in the cost of CPU and memory use.
+On a high-end laptop, computing the sets for all possible variations of the example above generates *8* sets and takes only *700 nanoseconds*, but computing the same for four sources and 16 resources (a scenario theoretically possible in Firefox with one language pack and Preferences UI for example) generates over *4 billion* sets and takes over *2 minutes*. 
+
+It becomes crucial to micro-optimize the algorithm and to use [Dynamic Programming](https://en.wikipedia.org/wiki/Dynamic_programming) and [Memoization](https://en.wikipedia.org/wiki/Memoization) techniques to minimize the cost of constructing, storing and retrieving sets.
```
Comment (dminor) [inline src/solver/README.md]: I think of dynamic programming more in terms of optimizing something by recursively solving smaller sub-problems, but this seems more like a combinatorial enumeration problem. I think the stuff you have below on backtracking explains things better, so it might make sense to move it here.

File: src/solver/README.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+Second important observation is that in most scenarios the resource exists in only some of the sources, and ability to bail out from a branch of candidates that cannot lead to a solution is crucial for performance of the algorithm.
+
+## Optimizations
+
+The algorithm used here is heavily optimized. For the conservative scenario listed above, where 4 sources and 15 resources are all present in every source, the total time on the reference hardware is cut from *2 minutes* to *24 seconds*, while generating the same *4 billion* sets for a **5x** performance improvement.
```
Comment (dminor) [inline src/solver/README.md]: I think "highly efficient" might be better than "heavily optimized" here, since this is an algorithmic improvement rather than hand-writing assembly or something like that.

File: src/solver/README.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+Second important observation is that in most scenarios the resource exists in only some of the sources, and ability to bail out from a branch of candidates that cannot lead to a solution is crucial for performance of the algorithm.
+
+## Optimizations
+
+The algorithm used here is heavily optimized. For the conservative scenario listed above, where 4 sources and 15 resources are all present in every source, the total time on the reference hardware is cut from *2 minutes* to *24 seconds*, while generating the same *4 billion* sets for a **5x** performance improvement. 
+
+### Streaming Iterator
+Regular `Iterator` in Rust requires the `Item` returned from `next` to be owned and not bound to the iterators lifetime. The concept of streaming iterator allows for `next` to return such reference which in this case is perfectly usable since all we need to return is a slice with indexes of sources for the `FluentBundle` to be generated from.
```
Comment (dminor) [inline src/solver/README.md]: I found this a little hard to follow. Maybe say something like:  "Unlike a regular iterator, a streaming iterator allows a borrowed reference to be returned..."

File: src/solver/README.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+┌┲━┱┬╂─╂┬───┐
+│┃0┃│┃1┃│   │
+└╂─╂┴┺━┹┴───┘
+ ┃ ┃    
+ ┗━┛ 
+ ```
+
+and that will potentially lead to a partial solution `[0, 1, ]` to be stored for the next iteration.
+
+If the test fails and no more sources can be generated, the algorithm will *backtrack* from to previous sources looking for a cell with the **highest** index prior to the cell that was being evaluated which is not yet on the last source. If such cell is found, the results of all cells **to the right** of the newfound cell are **erased** and the next branch can be evaluated.
```
Comment (dminor) [inline src/solver/README.md]: Does highest index here mean the source where you have the longest known potential solution?

nit: There is a missing word after *backtrack* from... I assume *backtrack* from the current cell.

File: src/solver/README.md [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+If no such cell can be found, that means that the iterator is complete.
+
+### Parallel
+
+If the testing can be performed in parallel, like an asynchronous I/O, the above *serial* solution is sub-optimal as it misses on the benefit of testing multiple cells at once.
+
+In such scenario, a difference to the algorithm is that it will not first populate a complete candidate such that all cells are at indexes that may lead to a solution and then call the testing method on such complete candidate:
```
Comment (dminor) [inline src/solver/README.md]: I find this a little confusing as well. I think what you are saying is that the partial solution is not necessarily valid, because you haven't tested every cell yet, because the tests are running in parallel.

File: src/solver/README.md [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+If the result a set of successes, the candidate is returned as a solution, and the algorithm proceeds to the same operation as if it was a failure.
+
+#### Failure case
+If the result contains failures, the iterator will now backtrack to find the closest lower or equal cell to the current index which can be advanced to the next source.
+In the example state above, the current cell can be advanced to *source 1* and then just a set of `[None, None, 1]` is to be evaluated by the tester (since we know that *A0* and *B0* are valid).
+
+If that is successful, the `[0, 0, 1]` set is a complete solution and is yielded.
+
+Then, the next state to be tested is:
```
Comment (dminor) [inline src/solver/README.md]: If we've found a complete solution, does the algorithm continue iterating? Or is this the failure case for the above?

File: src/solver/README.md [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+└╂─╂┴───┴───┘
+ ┃ ┃     
+ ┗━┛
+```
+###### Diagrams generated with use of http://marklodato.github.io/js-boxdrawing/
+
+where the horizontal block is a candidate, vertical block is a set of sources possible for each resource, and the arrow represents the index of a resource the iterator is currently evaluating.
+
+With those tools introduced, we can now guide the reader through how the algorithm works.
+But before we do that, it is important to justify writing a custom algorithm in place of existing generic solutions, and explain the two testing strategies which heavily impact the algorithm.
```
Comment (dminor) [inline src/solver/README.md:80]: I think part of the justification, that I missed the first time through, is that we care about the order in which we generate the possible combinations. We want to examine the likeliest combinations first, and then visit other possibilities only if the likely ones fail. An existing library wouldn't do this.

**Review by dminor — COMMENTED** [review]
An initial pass only looking at the readme. I'll try to get to the code later this week.

File: src/solver/README.md (hunk 7 above)
Comment (zbraniecki) [reply src/solver/README.md]: No, it means that if we are on a cell idx 3, we want to find the highest cell with index lower than 3 that is not yet "maxed out".

So, we'll see if cell idx 2 can have source bumped. If not, if cell idx 1 can etc. The first lower than 3 that we find is where we stop backtracking, because now we can bump it's source and move forward again.

How would you phrase such sentence? :)

File: src/solver/README.md (hunk 9 above)
Comment (zbraniecki) [reply src/solver/README.md]: It yields, and if it is resumed later, it will try to build the next complete solution.

File: src/solver/README.md (hunk 10 above)
Comment (zbraniecki) [reply src/solver/README.md:80]: I'm not sure if I understand. A regular multi cartesian product iterator would do the same - it would return the "highest value" result first assuming we order the input of sources properly.

In fact, prior to this PR, the library is using multi cartesian product, and the only issue with it is performance.

Comment (dminor) [reply src/solver/README.md:80]: Fair enough, I hadn't considering ordering the input sources.

File: src/solver/README.md (hunk 9 above)
Comment (dminor) [reply src/solver/README.md]: Ok, that makes sense. Maybe make this say "Then, if the iterator is resumed, the next state..."

File: src/solver/README.md (hunk 7 above)
Comment (dminor) [reply src/solver/README.md]: Maybe I understand better now: so if it fails, we backtrack from right to left on the partial solution, looking for a cell that is not "maxed out". If we find such a cell, we increment it, and "prune" everything to the right of it from the partial solution?

So [0, 0, 1, 1] would back track to [0, 0, ], increment to [0, 1], and get rid of everything to the right?

I was tripping up over cell index vs the cell contents and whether highest was the cell index or referred to a cell whose contents were the highest source.

Comment (zbraniecki) [reply src/solver/README.md]: Yeah! Maybe we should add it to the legend at the top?
I was also asked to consider generalizing it and releasing as a standalone and maybe white paper.
If so, we'd want to move away from "resources" to "keys" or some other term.
I think we can keep "source" for the other axis and "test" for the memoized operation.

File: src/solver/solution.rs [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            while self.is_cell_missing(i, source_idx) {
+                if source_idx >= self.depth - 1 {
+                    return false;
+                }
+                source_idx += 1;
+            }
+            // println!("prune new candidate: {}, {}", i, source_idx);
+            self.candidate[i] = source_idx;
+        }
+        // println!("prune: {:?}", self.candidate);
```
Comment (dminor) [inline src/solver/solution.rs]: nit: looks like leftover debugging here and at line 85. These should probably be removed or converted to logging.

File: src/solver/README.md (hunk 7 above)
Comment (dminor) [reply src/solver/README.md]: That make sense. I think using a word other than "resources" would help with readability, I found it difficult to keep track of which was which with "resource" and "source" just because of the similarity in the words.

Comment (zbraniecki) [reply src/solver/README.md]: I'll change that when I'll be looking to generalize the algo and extract it from this crate. For now I think it's ok to keep resources/sources.

File: examples/preferences.rs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+#[tokio::main]
+async fn main() {
+    let locales = vec!["en-US".parse().unwrap()];
+    let mut reg = L10nRegistry::default();
+
+    let browser_fs = l10nregistry::tokio::file_source(
+        "browser".to_string(),
+        locales.clone(),
+        "/home/zbraniecki/projects/l10nregistry-rs/tests/resources/browser/{locale}".into(),
```
Comment (dminor) [inline examples/preferences.rs]: Should this be a relative path?

File: examples/preferences.rs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    );
+    let toolkit_fs = l10nregistry::tokio::file_source(
+        "toolkit".to_string(),
+        locales.clone(),
+        "/home/zbraniecki/projects/l10nregistry-rs/tests/resources/toolkit/{locale}".into(),
```
Comment (dminor) [inline examples/preferences.rs]: Should this be a relative path?

File: src/registry/mod.rs [hunk 14]
```diff
@@ -13,57 +11,39 @@ use crate::source::FileSource;
...
-    length: usize,
-) -> itertools::MultiProduct<Rev<Range<usize>>> {
-    (0..length)
-        .map(|_| (0..source_count).rev())
-        .multi_cartesian_product()
-}
-
 #[derive(Default)]
-struct Shared {
+pub struct Shared {
```
Comment (dminor) [inline src/registry/mod.rs]: If this is going to be public, a more descriptive name might be good.

File: src/registry/mod.rs [hunk 15]
```diff
@@ -13,57 +11,39 @@ use crate::source::FileSource;
...
-    sources: RefCell<ChunkyVec<FileSource>>,
+    pub sources: RefCell<ChunkyVec<FileSource>>,
 }
 
 pub struct L10nRegistryLocked<'a> {
     lock: Ref<'a, ChunkyVec<FileSource>>,
 }
 
 impl<'a> L10nRegistryLocked<'a> {
+    /* ? */
```
Comment (dminor) [inline src/registry/mod.rs]: nit: update comment :)

File: src/registry/synchronous.rs [hunk 16]
```diff
@@ -1,36 +1,19 @@
...
-            } else {
-                return None;
-            }
-        }
-        Some(result)
+        source: usize,
+        res_id: &str,
+    ) -> Option<Rc<FluentResource>> {
+        let source = self.source_idx(source);
+        source.fetch_file_sync(langid, res_id)
```
Comment (dminor) [inline src/registry/synchronous.rs]: I think this might be more readable combined on one line so you can avoid introducing the new 'source' variable.

File: examples/async.rs [hunk 17]
```diff
@@ -0,0 +1,21 @@
+use l10nregistry::solver::testing::get_scenarios;
+use l10nregistry::solver::ParallelProblemSolver;
+
+#[tokio::main]
+async fn main() {
+    let scenarios = get_scenarios();
+
+    let scenario = scenarios
```
Comment (dminor) [inline examples/async.rs:8]: Please add a comment explaining why this scenario was chosen.

File: examples/sync.rs [hunk 18]
```diff
@@ -0,0 +1,16 @@
+use l10nregistry::solver::testing::get_scenarios;
+use l10nregistry::solver::SerialProblemSolver;
+
+fn main() {
+    let scenarios = get_scenarios();
+
+    let scenario = scenarios.iter().find(|s| s.name == "incomplete").unwrap();
```
Comment (dminor) [inline examples/sync.rs]: Please add a comment explaining why this scenario was chosen.

File: src/solver/solution.rs [hunk 19]
```diff
@@ -0,0 +1,115 @@
+use crate::fluent::FluentResource;
+use std::rc::Rc;
+
+#[derive(Debug)]
+pub struct Solution {
+    pub width: usize,
+    pub depth: usize,
+    pub candidate: Vec<usize>,
+    pub idx: usize,
+    pub dirty: bool,
```
Comment (dminor) [inline src/solver/solution.rs:10]: I think something like "invalidated" would be more readable than "dirty" here.

File: src/solver/solution.rs [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+            self.candidate[i] = source_idx;
+        }
+        true
+    }
+
+    pub fn is_complete(&self) -> bool {
+        self.idx == self.width - 1
+    }
+
+    pub fn bail(&mut self) -> bool {
```
Comment (dminor) [inline src/solver/solution.rs:99]: I think "try_advance_or_backtrack" or something similar would be more descriptive.

File: src/solver/mod.rs [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+
+pub use parallel::ParallelProblemSolver;
+pub use serial::SerialProblemSolver;
+pub use solution::Solution;
+
+use crate::fluent::FluentBundle;
+use crate::registry::L10nRegistry;
+use unic_langid::LanguageIdentifier;
+
+pub struct ProblemSolver {
```
Comment (dminor) [inline src/solver/mod.rs:14]: "ProblemSolver" is very generic. I'm terrible at naming things, so I won't argue if you want to keep this, but I think calling this something like "BundleResourceSolver" would give someone just scanning the code a better idea of what is going on.

I have more comments on naming below. We're using generic names in places, but the code itself has a lot of l10n specific bits. I think we should either completely abstract the problem solving bit away from the l10n stuff, or use names that are more descriptive for l10n. My vote would be to use more descriptive names.

File: src/tokio.rs [hunk 22]
```diff
@@ -16,7 +16,7 @@ impl FileFetcher for TokioFileFetcher {
         let mut f = ::tokio::fs::File::open(path).await?;
         let mut s = String::new();
         f.read_to_string(&mut s).await?;
-        println!("read = {}", s);
+        // println!("read = {}", s);
```
Comment (dminor) [inline src/tokio.rs]: nit: leftover debugging, please remove or making into logging

File: src/solver/solution.rs [hunk 23]
```diff
@@ -0,0 +1,115 @@
+use crate::fluent::FluentResource;
+use std::rc::Rc;
+
+#[derive(Debug)]
+pub struct Solution {
+    pub width: usize,
```
Comment (dminor) [inline src/solver/solution.rs]: Is "width" the number of resources? I'd suggest "(num|total|max)_resources"

File: src/solver/solution.rs [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+    pub depth: usize,
```
Comment (dminor) [inline src/solver/solution.rs:7]: Is "depth" the number of sources? I'd suggest "(num|total|max)_sources".

File: src/solver/solution.rs [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+    pub candidate: Vec<usize>,
+    pub idx: usize,
```
Comment (dminor) [inline src/solver/solution.rs]: Is "idx" the current resource we're looking at in the candidate? I'd suggest "cur_resource".

File: src/solver/solution.rs [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+    pub dirty: bool,
+
+    pub cache: Vec<Vec<Option<Option<Rc<FluentResource>>>>>,
```
Comment (dminor) [inline src/solver/solution.rs:17]: I'm having difficulty wrapping my head around the nested Options here. Please add a comment :)

**Review by dminor — COMMENTED** [review]
I didn't feel comfortable reviewing the ParallelSolver, but the rest is looking good to me!

File: examples/async.rs (hunk 17 above)
Comment (zbraniecki) [reply examples/async.rs:8]: I don't think we chose any particular scenario. As I'm testing the code, I'm changing which scenario we're examplifying here. I'd like to eventually just make it a command line argument.

File: src/solver/solution.rs (hunk 19 above)
Comment (zbraniecki) [reply src/solver/solution.rs:10]: Why `invalidated`? It was successfully returned, so we need to progress. I agree that `dirty` is a weird name, but I don't think `invalidated` communicates the state better. Other ideas?

File: src/solver/mod.rs (hunk 21 above)
Comment (zbraniecki) [reply src/solver/mod.rs:14]: My direction is to move all of the names to generic and extract the code into separate crate after I'm done with what's needed for Gecko.
I'd prefer not to start introducing less generic names, and Problem Solver is the technical name of the class of those algorithms - https://www.tannrallard.tech/intuitive-backtracking-algorithms/

I'm going to keep generic names and move toward them. Once we extract the crate, we will use more specific names in the callees here, and generic in the extracted crate.

File: src/solver/solution.rs (hunk 23 above)
Comment (zbraniecki) [reply src/solver/solution.rs]: I went with `res_len` to match `res_idx`.

File: src/solver/solution.rs (hunk 25 above)
Comment (zbraniecki) [reply src/solver/solution.rs]: I went with `res_idx` as I used it already in functions.

**Comment by zbraniecki** [comment]
This has been merged using a newer pruning solver branch!
